File: sharedmodule/llmswitch-core/src/router/virtual-router/engine-selection/tier-load-balancing.ts
import type { LoadBalancingPolicy, RoutePoolLoadBalancingPolicy, RoutePoolTier } from '../types.js';
import type { ProviderRegistry } from '../provider-registry.js';
import { extractProviderId, getProviderModelId } from './key-parsing.js';

export type ResolvedTierLoadBalancing = {
  strategy: LoadBalancingPolicy['strategy'];
  weights?: Record<string, number>;
};

export function resolveTierLoadBalancing(
  tier: RoutePoolTier,
  globalPolicy?: LoadBalancingPolicy
): ResolvedTierLoadBalancing {
  const tierPolicy = tier.loadBalancing;
  return {
    strategy: tierPolicy?.strategy ?? globalPolicy?.strategy ?? 'round-robin',
    weights: tierPolicy?.weights ?? globalPolicy?.weights
  };
}

export function resolveGroupWeight(groupId: string, weights?: Record<string, number>): number {
  if (!weights) {
    return 1;
  }
  const direct = weights[groupId];
  if (typeof direct === 'number' && Number.isFinite(direct) && direct > 0) {
    return direct;
  }
  const providerId = groupId.split('.')[0] ?? groupId;
  const providerOnly = weights[providerId];
  if (typeof providerOnly === 'number' && Number.isFinite(providerOnly) && providerOnly > 0) {
    return providerOnly;
  }
  return 1;
}

export function buildGroupWeights(
  groups: Map<string, string[]>,
  weights?: Record<string, number>
): Record<string, number> | undefined {
  if (!groups.size || !weights) {
    return undefined;
  }
  const out: Record<string, number> = {};
  let hasExplicit = false;
  for (const [groupId] of groups.entries()) {
    const resolved = resolveGroupWeight(groupId, weights);
    out[groupId] = resolved;
    if (resolved !== 1) {
      hasExplicit = true;
    }
  }
  return hasExplicit ? out : undefined;
}

export function hasNonUniformWeights(candidates: string[], weights?: Record<string, number>): boolean {
  if (!weights || candidates.length < 2) {
    return false;
  }
  let ref: number | undefined;
  for (const key of candidates) {
    const raw = weights[key];
    if (typeof raw !== 'number' || !Number.isFinite(raw)) {
      continue;
    }
    if (ref === undefined) {
      ref = raw;
    } else if (Math.abs(raw - ref) > 1e-6) {
      return true;
    }
  }
  return false;
}

export function buildCandidateWeights(opts: {
  candidates: string[];
  providerRegistry: ProviderRegistry;
  staticWeights?: Record<string, number>;
  dynamicWeights?: Record<string, number>;
}): Record<string, number> | undefined {
  const { candidates, providerRegistry, staticWeights, dynamicWeights } = opts;
  if ((!staticWeights || Object.keys(staticWeights).length === 0) && (!dynamicWeights || Object.keys(dynamicWeights).length === 0)) {
    return undefined;
  }

  const out: Record<string, number> = {};
  let hasExplicit = false;

  for (const key of candidates) {
    const dynamic = dynamicWeights?.[key];
    const staticWeight = resolveCandidateWeight(key, staticWeights, providerRegistry);
    const resolved = multiplyPositiveWeights(dynamic, staticWeight);
    if (resolved !== undefined) {
      out[key] = resolved;
      if (resolved !== 1) {
        hasExplicit = true;
      }
    }
  }

  if (!hasExplicit) {
    return undefined;
  }
  return out;
}

function resolveCandidateWeight(
  key: string,
  weights: Record<string, number> | undefined,
  providerRegistry: ProviderRegistry
): number | undefined {
  if (!weights) {
    return undefined;
  }
  const direct = normalizePositiveWeight(weights[key]);
  if (direct !== undefined) {
    return direct;
  }
  const providerId = extractProviderId(key) ?? '';
  if (!providerId) {
    return undefined;
  }
  try {
    const modelId = getProviderModelId(key, providerRegistry) ?? '';
    if (modelId) {
      const grouped = normalizePositiveWeight(weights[`${providerId}.${modelId}`]);
      if (grouped !== undefined) {
        return grouped;
      }
    }
  } catch {
    // Ignore registry misses and fall back to provider-only weight.
  }
  return normalizePositiveWeight(weights[providerId]);
}

function normalizePositiveWeight(value: number | undefined): number | undefined {
  return typeof value === 'number' && Number.isFinite(value) && value > 0 ? value : undefined;
}

function multiplyPositiveWeights(...values: Array<number | undefined>): number | undefined {
  let resolved: number | undefined;
  for (const value of values) {
    const normalized = normalizePositiveWeight(value);
    if (normalized === undefined) {
      continue;
    }
    resolved = resolved === undefined ? normalized : Math.max(1, Math.round(resolved * normalized));
  }
  return resolved;
}

File: sharedmodule/llmswitch-core/rust-core/crates/router-hotpath-napi/src/virtual_router_engine/engine/tier_load_balancing.rs
use std::collections::HashMap;

use crate::virtual_router_engine::load_balancer::LoadBalancingPolicy;
use crate::virtual_router_engine::routing::{extract_provider_id, RoutePoolTier};

#[derive(Debug, Clone)]
pub(crate) struct ResolvedTierLoadBalancing {
    pub strategy: String,
    pub weights: Option<HashMap<String, i64>>,
}

pub(crate) fn resolve_tier_load_balancing(
    tier: &RoutePoolTier,
    global_policy: &LoadBalancingPolicy,
) -> ResolvedTierLoadBalancing {
    let strategy = tier
        .load_balancing
        .as_ref()
        .and_then(|cfg| cfg.strategy.clone())
        .or_else(|| global_policy.strategy.clone())
        .unwrap_or_else(|| "round-robin".to_string());
    let weights = tier
        .load_balancing
        .as_ref()
        .and_then(|cfg| cfg.weights.clone())
        .or_else(|| global_policy.weights.clone());
    ResolvedTierLoadBalancing { strategy, weights }
}

pub(crate) fn build_group_weights(
    groups: &HashMap<String, Vec<String>>,
    weights: Option<&HashMap<String, i64>>,
) -> Option<HashMap<String, i64>> {
    let weights = match weights {
        Some(weights) => weights,
        None => return None,
    };
    let mut out = HashMap::new();
    let mut has_explicit = false;
    for group_id in groups.keys() {
        let weight = resolve_group_weight(group_id, Some(weights));
        if weight != 1 {
            has_explicit = true;
        }
        out.insert(group_id.clone(), weight);
    }
    if has_explicit {
        Some(out)
    } else {
        None
    }
}

fn resolve_group_weight(group_id: &str, weights: Option<&HashMap<String, i64>>) -> i64 {
    let weights = match weights {
        Some(weights) => weights,
        None => return 1,
    };
    if let Some(weight) = normalize_positive_weight(weights.get(group_id).cloned()) {
        return weight;
    }
    let provider_id = extract_provider_id(group_id).unwrap_or_else(|| group_id.to_string());
    normalize_positive_weight(weights.get(provider_id.as_str()).cloned()).unwrap_or(1)
}

fn normalize_positive_weight(value: Option<i64>) -> Option<i64> {
    value.filter(|weight| *weight > 0)
}
